### PR TITLE
Add new API reference style and extension plugin guide

### PR DESCRIPTION
**Pull request by JTeijema**
So, this in a new implementation of the API reference, formatted in a tree structure as opposed to a list. It also contains an updates extension created guide.

Yet to do is a hunt for dead links, but I wanted to start the pull request for review.

Note: the API reference can be automated using the :recursive: trigger in the autosummary. I've tested this and it looks very promissing, but I wanted to get this implemented first.

### CODE REVIEW

**Comment by JTeijema** [comment]
and also to fix the test errors apparently

**Comment by ottomattas** [comment]
Good work!

I love the tree structure while I am not convinced the extension documentation should be a branch from API.
Just a question to get some more insight on your thought process - have you considered splitting extensions away from the API logically? I mean, instead of `docs/source/API/extensions_dev.rst` + `docs/source/API/reference.rst`, you would instead do something like `docs/source/extensions/development.rst` + `docs/source/API/reference.rst`.

(Why I think this is worth considering: we can then simply add additional docs without worrying about logical confusion. Think FAQs, for example. `docs/source/API/extensions_dev/FAQ.rst` + `docs/source/API/FAQ.rst` vs `docs/source/extensions/FAQ.rst` + `docs/source/API/FAQ.rst`)

**Comment by JTeijema** [comment]
@ottomattas Are we looking at the same version? I don't think docs/source/API/extensions_dev.rs is part of this PR anymore hahaha. I came to the same conclusion as you :)

**Comment by ottomattas** [comment]
You are right @JTeijema!
I saw there was a huge diff for those deprecated files which I did not open. Actually the diff was that they are deleted.

**Comment by JTeijema** [comment]
And yeah, I totally agree with the extension being out of place in the API hahaha.

File: docs/source/extensions/extension_development.rst [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (J535D165) [inline docs/source/extensions/extension_development.rst:97]:
```suggestion
```

Please remove the additional linebreaks.

File: docs/source/extensions/extension_development.rst [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+Dataset extensions
+------------------
+An extension of the :class:`asreview.datasets.BaseDataSet` class.
+
+Dataset extensions integrate new datasets for use in ASReview. Adding datasets
+via extension provides quick access to the dataset via commandline or other user
+interface.
```
Comment (J535D165) [inline docs/source/extensions/extension_development.rst:186]:
```suggestion
Dataset extensions integrate new datasets for use in ASReview. Adding datasets
via extension provides quick access to the dataset via Command Line Interface or in
ASReview LAB.
```

Let's add cross-references to the relevant parts of CLI and LAB.

File: docs/source/extensions/extension_development.rst [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+.. code:: bash
+
+    ├── README.md
+    ├── asreviewcontrib
+    │   └── dataset_name
+    │       ├── __init__.py
+    │       └── your_dataset.py
+    ├── config
+    │   └── your_dataset.json
```
Comment (J535D165) [inline docs/source/extensions/extension_development.rst:199]: I would suggest not use the config option for this example. This is pretty advanced usage. This was used in ASReview-covid to make automated updates available. Most use cases don't need this I think.

File: docs/source/extensions/extension_development.rst [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    ├── datasets
+    │   └── dataset_name
+    │       ├── __init__.py
```
Comment (J535D165) [inline docs/source/extensions/extension_development.rst:202]: No need to have an `__init__.py` file in a data folder.

File: docs/source/extensions/extension_development.rst [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+    ├── README.md
+    ├── asreviewcontrib
+    │   └── dataset_name
+    │       ├── __init__.py
+    │       └── your_dataset.py
+    ├── config
+    │   └── your_dataset.json
+    ├── datasets
+    │   └── dataset_name
```
Comment (J535D165) [inline docs/source/extensions/extension_development.rst:201]:
```suggestion
    ├── data
```

File: docs/source/extensions/extension_development.rst [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    │       ├── __init__.py
+    │       └── your_dataset.csv
+    ├── setup.py
+    └── tests
+
```
Comment (J535D165) [inline docs/source/extensions/extension_development.rst:206]: I think we need an example of `your_dataset.py` here. Covid plugin is relatively complicated.

I think a minimal example would look like this:

```
from asreview.datasets import BaseDataSet
from asreview.datasets import BaseDataGroup

class YourDataGroup(BaseDataGroup):
    group_id = "your_data_group"
    description = "A new data group with my awesome datasets."

    def __init__(self):
        
        dataset = BaseDataSet.from_config({
            "dataset_id": "your_data_id",
            "url": "",
            "reference": "",
            "link": "",
            "license": "",
            "title": "Your Data",
            "authors": [
              "Jane Doe",
              "John Doe"
            ],
            "year": 2021,
            "topic": "Your topic",
            "final_inclusions": true,
            "title_abstract_inclusions": false
          }
        )

        super(YourDataGroup, self).__init__(dataset)
        # pass multiple datasets to init if there are more datasets
```

File: docs/source/extensions/overview_extensions.rst [hunk 7]
```diff
@@ -72,3 +69,28 @@ The following extensions are developed and maintained by the ASReview community:
 If an extension is not on this list, or you made one and you would like it to 
 be added to this list, please initiate an issue on `Github
 <https://github.com/asreview/asreview/issues>`__.
+
+
+Usage
+-----
```
Comment (J535D165) [inline docs/source/extensions/overview_extensions.rst:75]: "Installation" would fit the context.

File: docs/source/index.rst [hunk 8]
```diff
@@ -66,6 +66,8 @@ A (citable) PDF of the ASReview documentation can be found on `Zenodo <https://d
 
    extensions/overview_extensions
 
+   extensions/extension_development
```
Comment (J535D165) [inline docs/source/index.rst:69]: @Rensvandeschoot I need your help here. I think this belongs to the development section (see below).

File: docs/source/extensions/extension_development.rst [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+For minimal functionality, ``your_dataset.py`` should extent
+:class:`asreview.datasets.BaseDataSet`. Further functionality can be extensions
+of any other class in :mod:`asreview.datasets`.
```
Comment (J535D165) [inline docs/source/extensions/extension_development.rst:209]: I would suggest creating a template repo here. Like https://github.com/asreview/template-extension-new-model

This should be straightforward with the example above.

File: docs/source/extensions/extension_development.rst [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    │   └── your_dataset.json
+    ├── datasets
+    │   └── dataset_name
+    │       ├── __init__.py
+    │       └── your_dataset.csv
+    ├── setup.py
+    └── tests
+
+For minimal functionality, ``your_dataset.py`` should extent
+:class:`asreview.datasets.BaseDataSet`. Further functionality can be extensions
```
Comment (J535D165) [inline docs/source/extensions/extension_development.rst:208]: Maybe also explain what `asreview.datasets.BaseDataGroup` does?

**Review by J535D165 — CHANGES_REQUESTED** [review]
Thanks a lot for this PR. I love the new reference style. Relatively minor feedback to address.

**Comment by J535D165** [comment]
Any idea why the tests fail?

File: docs/source/index.rst (hunk 8 above)
Comment (JTeijema) [reply docs/source/index.rst:69]: I've moved it to development. (Also to manage broken links)

**Comment by JTeijema** [comment]
As there is only one documented exception, I am removing the exception block from the template:
```
   {% block exceptions %}
   {% if exceptions %}
   .. rubric:: {{ _('Exceptions') }}

   .. autosummary::
      :toctree:
   {% for item in exceptions %}
      {{ item }}
   {%- endfor %}
   {% endif %}
   {% endblock %}
```

The only exception listed by the API reference was: `asreview.datasets.DataSetNotFoundError`

**Comment by JTeijema** [comment]
Do we want with or without signatures?
afbeelding

afbeelding


I'm thinking without. 

Also, the error is within the template, so I guess I'm learning Jinja

**Comment by JTeijema** [comment]
See: 
afbeelding

**Comment by JTeijema** [comment]
Yes ofcourse because module calls class. ^

**Comment by JTeijema** [comment]
Okay so both classes and func triggers the problem

**Comment by JTeijema** [comment]
Oh c'mon. It cant be both those things. Those are totally unrelated things.

**Comment by JTeijema** [comment]
THIS IS THE SAME AS BEFORE WHY ARE YOU BUILDING?

Anyway, test's fixed :)

**Comment by J535D165** [comment]
Thanks!!

File: docs/source/API/extension_dev.rst [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+example`` subcommand. See ``asreview -h`` for this option.
+
+An example of a subcommand extension is the `Visualization Extension
+<https://github.com/asreview/asreview-visualization>`_
+
+Dataset extensions
+------------------
+An extension of the :class:`asreview.datasets.BaseDataSet` class.
+
+Dataset extensions integrate new datasets for use in ASReview. Adding datasets
```
Comment (Rensvandeschoot) [inline docs/source/API/extension_dev.rst:164]: Can you maybe add an explanation how the 'dataset' as used here, is different from a dataset you can import or a benchmark dataset you can see in the exploration mode?

**Pull request merged**